src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -404,114 +383,93 @@ async def build_pex(
-            metadata = PythonLockfileMetadata.from_lockfile(
-                lock_bytes,
-                **(dict() if synthetic_lock else dict(lockfile_path=lock_path)),
-                resolve_name=resolve_name,
-                delimeter=header_delimiter,
-            )
-
+            argv.extend(["--requirement", lockfile.lockfile_path, "--no-transitive"])
+            is_network_resolve = True
+        if lockfile.metadata and request.requirements.complete_req_strings:
             validate_metadata(
-                metadata, request.interpreter_constraints, request.requirements, python_setup  # type: ignore[arg-type]
+                lockfile.metadata,
+                request.interpreter_constraints,
+                lockfile.original_lockfile,
+                request.requirements.complete_req_strings,
+                python_setup,
             )
-
     else:
-        assert isinstance(request.requirements, PexRequirements)
-        is_monolithic_resolve = request.requirements.is_all_constraints_resolve
-        requirement_count = len(request.requirements.req_strings)
-
-        if request.requirements.constraints_strings:
-            constraints_file = "__constraints.txt"
-            constaints_content = "\n".join(request.requirements.constraints_strings)
-            constraints_file_digest = await Get(
-                Digest,
-                CreateDigest([FileContent(constraints_file, constaints_content.encode())]),
-            )
-            argv.extend(["--constraints", constraints_file])
-
+        # TODO: This is not the best heuristic for available concurrency, since the
+        # requirements almost certainly have transitive deps which also need building, but it
+        # is better than using something hardcoded.
+        concurrency_available = len(request.requirements.req_strings)
         argv.extend(request.requirements.req_strings)
 
+        if isinstance(request.requirements.from_superset, Pex):
+            repository_pex = request.requirements.from_superset
+            argv.extend(["--pex-repository", repository_pex.name])
+            requirements_digests.append(repository_pex.digest)
+            is_network_resolve = False
+        elif isinstance(request.requirements.from_superset, LoadedLockfile):
+            loaded_lockfile = request.requirements.from_superset
+            # NB: This is also validated in the constructor.
+            assert loaded_lockfile.is_pex_native
+            requirements_digests.append(loaded_lockfile.lockfile_digest)
+            argv.extend(["--lock", loaded_lockfile.lockfile_path])
+            is_network_resolve = False
+
+            if loaded_lockfile.metadata:
+                validate_metadata(
+                    loaded_lockfile.metadata,
+                    request.interpreter_constraints,
+                    loaded_lockfile.original_lockfile,
+                    request.requirements.req_strings,
+                    python_setup,
+                )
```

Comment:
I'm having a hard time following every code path. Do you know if this will run for both tool like files and user locked files?

Reply:
It will run if `pex_requirements.should_validate`: https://github.com/pantsbuild/pants/pull/15014/files#diff-1cf75cd876e2d66cbfebc9e53e2d34a117424a44f49c039cbac0e26ae7c40861R176-R183